src/components/ContactImportation/ContactImportationModal.jsx

```diff
@@ -0,0 +1,133 @@
+import React from "react";
+import PropTypes from "prop-types";
+import {
+  Alerter,
+  Button,
+  Modal,
+  ModalContent,
+  ModalHeader
+} from "cozy-ui/react";
+import _ from "lodash";
+import { withContactsMutations } from "../../connections/allContacts";
+import ContactImportationCard from "./ContactImportationCard";
+import Importation from "../../importation";
+import Status from "../../importation/status";
+
+const ERROR_STATUS_SET = new Set([Status.FILE_ISSUE, Status.COMPLETE_FAILURE]);
+
+class ContactImportationModal extends React.Component {
+  state = {
+    progress: null
+  };
+
+  selectFile = file => {
+    this.props.updateImportation(
+      Importation.selectFile(file, this.props.importation)
+    );
+  };
+
+  unselectFile = () => {
+    this.props.updateImportation(
+      Importation.unselectFile(this.props.importation)
+    );
+  };
+
+  onProgress = progress => {
+    this.setState({ progress });
```

Comment:
`setState` is an asynchronous function. At runtime the callstack could be:
```js
onProgress(10);
onProgress(20);
onProgress(30);
onProgress(40);
onProgress(50);
onProgress(60);
onProgress(70);
onProgress(80);
onProgress(90);
onProgress(100);
```

which will be fine, but it could also be:

```js
onProgress(10);
onProgress(60);
onProgress(40);
onProgress(20);
onProgress(30);
onProgress(100);
onProgress(70);
onProgress(50);
onProgress(90);
onProgress(80);
```

You must use the functionnal `setState` with a verification on previous state value:

```js
onProgress = progress => {
  this.setState(state => ({...state, progress > state.progress ? progress : state.progress }))
}
```

Reply:
> setState is an asynchronous function.

Yes. But the order of the updates is preserved. Source : https://stackoverflow.com/questions/48563650/does-react-keep-the-order-for-state-updates?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa (by Dan Abramov no less ;) )

No need for the functional version of setState here :)

Reply:
TIL, I thought order wasn't necessarily preserved.
see https://stackoverflow.com/a/48610973/178575